### PR TITLE
tazeek/handle_sanction_failures

### PR DESCRIPTION
**Pull request by tazeek**

### CODE REVIEW

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 1]
```diff
@@ -75,21 +75,27 @@ sub _hmt_csv {
     open $fh, '+>', undef or die "Could not open anonymous temp file - $!";                    ## no critic (RequireBriefOpen)
     print $fh $content;
     seek($fh, 0, 0);
-    my $info = $csv->getline($fh);
 
-    while (my $row = $csv->getline($fh) or not $csv->eof) {
+    my $sanction_list;
+
+    while (my $row = $csv->getline($fh)) {
+        $sanction_list //= $row;
```
Comment (aminmarashi) [inline lib/Data/Validate/Sanctions/Fetcher.pm]: You can use `$last_update` instead and assign it like this:

```
$last_update //= $row->[1];
```

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 2]
```diff
@@ -75,21 +75,27 @@ sub _hmt_csv {
...
         ($row->[23] and $row->[23] eq "Individual") or next;
         my $name = _process_name @{$row}[0 .. 5];
         next if $name =~ /^\s*$/;
         push @names, $name;
     }
+
+    die "Sanction file is not accessible." unless $csv->eof;
```
Comment (aminmarashi) [inline lib/Data/Validate/Sanctions/Fetcher.pm]: I think at this point we can't be very specific about the error ourselves, so we can get help:

```
die "Getting HMT sancations failed: " . $csv->error_diag();
```

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 3]
```diff
@@ -75,21 +75,28 @@ sub _hmt_csv {
     open $fh, '+>', undef or die "Could not open anonymous temp file - $!";                    ## no critic (RequireBriefOpen)
     print $fh $content;
     seek($fh, 0, 0);
-    my $info = $csv->getline($fh);
 
-    while (my $row = $csv->getline($fh) or not $csv->eof) {
+    # Variable used to store the date of the last row
```
Comment (aminmarashi) [inline lib/Data/Validate/Sanctions/Fetcher.pm]: A little rephrase could be done here

```
# Shows the last date the sanctions list has been updated
```

Comment (aminmarashi) [reply lib/Data/Validate/Sanctions/Fetcher.pm]: Useful to mention this is retrieved from the first line of the csv file.

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 4]
```diff
@@ -61,6 +61,13 @@ sub _ofac_xml {
         pattern  => '%m/%d/%Y',
         on_error => 'croak',
     );
+
+    # Pattern to look for: mm/dd/yyy
+    my $pattern = '((0[1-9]|1[012])[/]0[1-9]|[12][0-9]|3[01])[/]((19|20)\d\d)';
+    if ($ref->{publshInformation}{Publish_Date} !~ m/$pattern/) {
+        die 'Date does not match the pattern of %m/%d/%Y';
```
Comment (arunbinary) [inline lib/Data/Validate/Sanctions/Fetcher.pm]: would be better to die with `Date does not match the pattern of mm/dd/yyyy` for better readability.

Comment (tazeek) [reply lib/Data/Validate/Sanctions/Fetcher.pm]: Done. Made changes for `_hmt_csv` as well.

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 5]
```diff
@@ -61,6 +61,13 @@ sub _ofac_xml {
         pattern  => '%m/%d/%Y',
         on_error => 'croak',
     );
+
+    # Pattern to look for: mm/dd/yyy
+    my $pattern = '((0[1-9]|1[012])[/]0[1-9]|[12][0-9]|3[01])[/]((19|20)\d\d)';
```
Comment (raunakkathuria) [inline lib/Data/Validate/Sanctions/Fetcher.pm]: better to move this to sub as its used here as well https://github.com/binary-com/perl-Data-Validate-Sanctions/pull/21/files#diff-f22d1e5548dc0825f328a4d72bbbcb96R90

Comment (tazeek) [reply lib/Data/Validate/Sanctions/Fetcher.pm]: The regex for each case is different: one checks for `dd/mm/yyyy` and the other for `mm/dd/yyyy`

Comment (basiliscos) [reply lib/Data/Validate/Sanctions/Fetcher.pm]: I think this is overcomplicated. The pattern should be like `/(\d){1,2}/(\d){1,2}/(\d){4}/` 

for `dd/mm/yyyy` then you unpack the match to `($d, $m, $y) = ($1, $2, $3)` and for `mm/dd/yyyy` you just unpack by `($d, $m, $y) = ($2, $1,  $3)`, and only then it should try to instantiate `Date` to make sure it is correct.

The complex patterns do not provide needed defence agains bad dates, i.e. 30 of February, or 31 of November.

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 6]
```diff
@@ -48,6 +48,18 @@ sub _ofac_xml_zip {
     return _ofac_xml($output);
 }
 
+sub _validate_date {
+
+    my $file_date = shift;
+
+    # Date validation checker for both mm/dd/yyyy and dd/mm/yyyy
+    if ($file_date !~ /^(\d{1,2})\/(\d{1,2})\/(\d{4})$/) {
```
Comment (binarytom) [inline lib/Data/Validate/Sanctions/Fetcher.pm]: These are usually easier to read with different delimiters.

I'd normally write this as:

```
return 1 if $file_date =~ m{^(\d{1,2})/(\d{1,2})/(\d{4})$};
return 0;
```

**Pull request merged**